File: src/odometry_pkg/scripts/position_visualizer.py
#!/usr/bin/env python3
"""
Position Visualizer Node for RVIZ2

Subscribes to PositionData and publishes visualization messages:
- nav_msgs/Path: Trajectory of the robot
- geometry_msgs/PoseStamped: Current pose
- visualization_msgs/Marker: Robot mesh marker (mecanum robot STL)
"""

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, DurabilityPolicy
from odometry_interfaces_pkg.msg import PositionData
from nav_msgs.msg import Path
from geometry_msgs.msg import PoseStamped, Point, TransformStamped
from visualization_msgs.msg import Marker
from std_msgs.msg import ColorRGBA
from tf2_ros import TransformBroadcaster
import math


class PositionVisualizer(Node):
    def __init__(self):
        super().__init__('position_visualizer')
        
        # Declare parameters for mesh
        self.declare_parameter('use_mesh', True)
        self.declare_parameter('mesh_scale', 0.001)  # STL is in mm, convert to meters
        self.use_mesh = self.get_parameter('use_mesh').value
        self.mesh_scale = self.get_parameter('mesh_scale').value

        # Geometry matching the URDF (box chassis + four wheel meshes)
        self.base_size = (0.40, 0.20, 0.10)
        self.base_offset_z = 0.05  # URDF base_link visual origin
        self.wheel_mesh = 'package://odometry_pkg/meshes/wheel.stl'
        self.wheel_frames = ["wheel_fl", "wheel_fr", "wheel_rl", "wheel_rr"]
        # Per-wheel orientation and scale: mirror right-side wheels using yaw flip + negative X scale
        self.wheel_mesh_orientation = self.rpy_to_quaternion(1.5708, 0.0, 0.0)
        yaw_flip = self.rpy_to_quaternion(1.5708, 0.0, 6.2832)
        self.wheel_orientations = {
            "wheel_fl": self.wheel_mesh_orientation,
            "wheel_fr": yaw_flip,
            "wheel_rl": yaw_flip,
            "wheel_rr": self.wheel_mesh_orientation,
        }
        self.wheel_scales = {
            "wheel_fl": (self.mesh_scale, self.mesh_scale, self.mesh_scale),
            "wheel_fr": (-self.mesh_scale, self.mesh_scale, self.mesh_scale),
            "wheel_rl": (-self.mesh_scale, self.mesh_scale, self.mesh_scale),
            "wheel_rr": (self.mesh_scale, self.mesh_scale, self.mesh_scale),
        }
        # Offsets: flip side gets +X (red), normal side gets -X; Y/Z shared
        self.wheel_offset_y = self.declare_parameter('wheel_offset_y', 0.025).value
        self.wheel_offset_z = self.declare_parameter('wheel_offset_z', -0.1).value
        self.wheel_offset_map_x = {
            "wheel_fl": -0.1,
            "wheel_fr": 0.1,
            "wheel_rl": 0.1,
            "wheel_rr": -0.1,
        }
        
        # Subscriber
        self.subscription = self.create_subscription(
            PositionData,
            '/odometry/position_from_accel',  # Will be remapped to /odometry/position_from_wheels
            self.position_callback,
            10
        )
        
        # Publishers
        self.path_pub = self.create_publisher(Path, '/odometry/path', 10)
        self.pose_pub = self.create_publisher(PoseStamped, '/odometry/pose', 10)
        self.marker_pub = self.create_publisher(Marker, '/odometry/robot_marker', 10)
        
        # TF broadcaster for robot pose
        self.tf_broadcaster = TransformBroadcaster(self)

        # Track current pose so we can publish TF even before new data arrives
        self.current_x = 0.0
        self.current_y = 0.0
        self.current_alpha = 0.0
        self.have_position = False

        # Keep TF alive even when no new PositionData is published
        self.tf_timer = self.create_timer(0.05, self.publish_last_tf)
        
        # Path message (accumulates poses)
        self.path_msg = Path()
        self.path_msg.header.frame_id = 'map'
        
        # Track last position for reset detection
        self.last_x = None
        self.last_y = None
        
        self.get_logger().info('Position Visualizer started')
        self.get_logger().info(f'Using mesh: {self.use_mesh}, scale: {self.mesh_scale}')
        self.get_logger().info('Subscribing to: /odometry/position_from_accel (remapped)')
        self.get_logger().info('Publishing: /odometry/path, /odometry/pose, /odometry/robot_marker')
        # Publish once immediately so all wheels appear without waiting for the first timer tick
        self.publish_last_tf()
    
    def position_callback(self, msg: PositionData):
        current_time = self.get_clock().now().to_msg()
        
        # Detect position reset (large jump back to origin)
        if self.last_x is not None and self.last_y is not None:
            dx = abs(msg.x - self.last_x)
            dy = abs(msg.y - self.last_y)
            # If position jumped more than 0.5m and is near origin, clear path
            if (dx > 0.5 or dy > 0.5) and abs(msg.x) < 0.1 and abs(msg.y) < 0.1:
                self.get_logger().info('Position reset detected - clearing path')
                self.path_msg.poses.clear()
        
        self.last_x = msg.x
        self.last_y = msg.y
        self.current_x = msg.x
        self.current_y = msg.y
        self.current_alpha = msg.alpha
        self.have_position = True
        
        # Create PoseStamped
        pose = PoseStamped()
        pose.header.stamp = current_time
        pose.header.frame_id = 'map'
        pose.pose.position.x = msg.x
        pose.pose.position.y = msg.y
        pose.pose.position.z = 0.0
        
        # Convert alpha (yaw) to quaternion
        yaw = msg.alpha
        pose.pose.orientation.x = 0.0
        pose.pose.orientation.y = 0.0
        pose.pose.orientation.z = math.sin(yaw / 2.0)
        pose.pose.orientation.w = math.cos(yaw / 2.0)
        
        # Publish current pose
        self.pose_pub.publish(pose)
        
        # Broadcast TF transform for the robot
        self.broadcast_tf(msg, current_time)
        
        # Add to path and publish
        self.path_msg.header.stamp = current_time
        self.path_msg.poses.append(pose)
        
        # Limit path length to prevent memory issues
        max_path_length = 5000
        if len(self.path_msg.poses) > max_path_length:
            self.path_msg.poses = self.path_msg.poses[-max_path_length:]
        
        self.path_pub.publish(self.path_msg)
        
        # Create robot marker (URDF-matching multi-part mesh or arrow)
        self.publish_robot_marker(msg, current_time)

    def publish_last_tf(self):
        """Publish the most recent transform (or a zeroed one on startup)."""
        stamp = self.get_clock().now().to_msg()
        msg = PositionData()
        msg.x = self.current_x
        msg.y = self.current_y
        msg.alpha = self.current_alpha
        # Always broadcast so RVIZ has a valid frame; before data arrives this sits at origin
        self.broadcast_tf(msg, stamp)
        # Keep markers alive and ensure they spawn immediately
        self.publish_robot_marker(msg, stamp)

    def broadcast_tf(self, msg: PositionData, stamp):
        """Broadcast TF transform from map to base_link."""
        t = TransformStamped()
        t.header.stamp = stamp
        t.header.frame_id = 'map'
        t.child_frame_id = 'base_link'
        
        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0
        
        # Quaternion from yaw
        yaw = msg.alpha
        t.transform.rotation.x = 0.0
        t.transform.rotation.y = 0.0
        t.transform.rotation.z = math.sin(yaw / 2.0)
        t.transform.rotation.w = math.cos(yaw / 2.0)
        
        self.tf_broadcaster.sendTransform(t)
    
    def publish_robot_marker(self, msg: PositionData, stamp):
        """Publish robot visualization marker."""
        if self.use_mesh:
            # Base cube matching URDF chassis (frame = base_link so TF controls pose)
            base_marker = Marker()
            base_marker.header.stamp = stamp
            base_marker.header.frame_id = 'base_link'
            base_marker.ns = 'robot'
            base_marker.id = 0
            base_marker.type = Marker.CUBE
            base_marker.action = Marker.ADD
            base_marker.frame_locked = True  # follow TF updates without re-publishing
            # Lifetime 0 means RViz keeps latest until overwritten; reusing IDs avoids churn
            base_marker.pose.position.x = 0.0
            base_marker.pose.position.y = 0.0
            base_marker.pose.position.z = self.base_offset_z
            base_marker.pose.orientation.x, base_marker.pose.orientation.y, \
                base_marker.pose.orientation.z, base_marker.pose.orientation.w = self.rpy_to_quaternion(0.0, 0.0, 0.0)
            base_marker.scale.x, base_marker.scale.y, base_marker.scale.z = self.base_size
            base_marker.color.r = 1.0
            base_marker.color.g = 0.0
            base_marker.color.b = 0.0
            base_marker.color.a = 0.7
            self.marker_pub.publish(base_marker)

            # Wheels as individual mesh markers parented to wheel TF frames (pose & rotation come directly from TF)
            for idx, frame in enumerate(self.wheel_frames, start=1):
                wheel_marker = Marker()
                wheel_marker.header.stamp = stamp
                wheel_marker.header.frame_id = frame
                wheel_marker.ns = 'robot'
                wheel_marker.id = idx
                wheel_marker.type = Marker.MESH_RESOURCE
                wheel_marker.action = Marker.ADD
                wheel_marker.frame_locked = True  # bind to wheel link TF
                # Bind exactly to the wheel link frame (position/orientation from TF)
                wheel_marker.pose.position.x = self.wheel_offset_map_x.get(frame, -0.1)
                wheel_marker.pose.position.y = self.wheel_offset_y
                wheel_marker.pose.position.z = self.wheel_offset_z
                ori = self.wheel_orientations.get(frame, self.wheel_mesh_orientation)
                wheel_marker.pose.orientation.x = ori[0]
                wheel_marker.pose.orientation.y = ori[1]
                wheel_marker.pose.orientation.z = ori[2]
                wheel_marker.pose.orientation.w = ori[3]

                sx, sy, sz = self.wheel_scales.get(frame, (self.mesh_scale, self.mesh_scale, self.mesh_scale))
                wheel_marker.scale.x = sx
                wheel_marker.scale.y = sy
                wheel_marker.scale.z = sz
                wheel_marker.mesh_resource = self.wheel_mesh
                wheel_marker.mesh_use_embedded_materials = False
                wheel_marker.color.r = 0.1
                wheel_marker.color.g = 0.1
                wheel_marker.color.b = 0.1
                wheel_marker.color.a = 0.8
                self.marker_pub.publish(wheel_marker)
        else:
            marker = Marker()
            marker.header.stamp = stamp
            marker.header.frame_id = 'map'
            marker.ns = 'robot'
            marker.id = 0
            marker.action = Marker.ADD
            marker.type = Marker.ARROW
            
            yaw = msg.alpha
            marker.pose.position.x = msg.x
            marker.pose.position.y = msg.y
            marker.pose.position.z = 0.0
            marker.pose.orientation.x = 0.0
            marker.pose.orientation.y = 0.0
            marker.pose.orientation.z = math.sin(yaw / 2.0)
            marker.pose.orientation.w = math.cos(yaw / 2.0)
            
            # Arrow size
            marker.scale.x = 0.3  # Length
            marker.scale.y = 0.05  # Width
            marker.scale.z = 0.05  # Height
            
            # Red color
            marker.color.r = 1.0
            marker.color.g = 0.0
            marker.color.b = 0.0
            marker.color.a = 1.0
            self.marker_pub.publish(marker)

    @staticmethod
    def rpy_to_quaternion(roll: float, pitch: float, yaw: float):
        """Convert roll, pitch, yaw to quaternion (x, y, z, w)."""
        cr = math.cos(roll * 0.5)
        sr = math.sin(roll * 0.5)
        cp = math.cos(pitch * 0.5)
        sp = math.sin(pitch * 0.5)
        cy = math.cos(yaw * 0.5)
        sy = math.sin(yaw * 0.5)

        x = sr * cp * cy - cr * sp * sy
        y = cr * sp * cy + sr * cp * sy
        z = cr * cp * sy - sr * sp * cy
        w = cr * cp * cy + sr * sp * sy
        return x, y, z, w

    @staticmethod
    def multiply_quaternion(q1, q2):
        """Quaternion multiplication q = q1 * q2 (xyzw)."""
        x1, y1, z1, w1 = q1
        x2, y2, z2, w2 = q2
        x = w1 * x2 + x1 * w2 + y1 * z2 - z1 * y2
        y = w1 * y2 - x1 * z2 + y1 * w2 + z1 * x2
        z = w1 * z2 + x1 * y2 - y1 * x2 + z1 * w2
        w = w1 * w2 - x1 * x2 - y1 * y2 - z1 * z2
        return x, y, z, w


def main(args=None):
    rclpy.init(args=args)
    node = PositionVisualizer()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/odometry_pkg/src/position_visualizer.cpp
/**
 * @file position_visualizer.cpp
 * @brief Position Visualizer - Publishes TF and visualization markers for RViz
 *
 * This node subscribes to odometry position data and publishes:
 * - TF transforms (map -> base_link)
 * - RViz visualization markers (robot mesh + path)
 * - Path message for trajectory visualization
 *
 * @author Group g1
 * @date 2025
 */

#include <chrono>
#include <memory>
#include <cmath>
#include <string>
#include <vector>
#include <array>
#include <iomanip>
#include <sstream>

#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "nav_msgs/msg/path.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "visualization_msgs/msg/marker.hpp"
#include "visualization_msgs/msg/marker_array.hpp"
#include "std_msgs/msg/int32_multi_array.hpp"
#include "tf2_ros/transform_broadcaster.h"
#include "ament_index_cpp/get_package_share_directory.hpp"

using namespace std::chrono_literals;

/**
 * @class PositionVisualizer
 * @brief ROS2 node that visualizes odometry data in RViz
 */
class PositionVisualizer : public rclcpp::Node
{
public:
    PositionVisualizer() : Node("position_visualizer")
    {
        // Declare parameters
        this->declare_parameter<bool>("use_mesh", true);
        this->declare_parameter<double>("mesh_scale", 0.001);
        this->declare_parameter<std::string>("mesh_path", "");
        this->declare_parameter<double>("path_width", 0.02);
        this->declare_parameter<int>("max_path_points", 10000);
        this->declare_parameter<int>("total_segments", 5);

        // Get parameters
        use_mesh_ = this->get_parameter("use_mesh").as_bool();
        mesh_scale_ = this->get_parameter("mesh_scale").as_double();
        mesh_path_ = this->get_parameter("mesh_path").as_string();
        path_width_ = this->get_parameter("path_width").as_double();
        max_path_points_ = this->get_parameter("max_path_points").as_int();
        total_segments_ = this->get_parameter("total_segments").as_int();

        // Default mesh path
        if (mesh_path_.empty()) {
            try {
                std::string pkg_share = ament_index_cpp::get_package_share_directory("odometry_pkg");
                mesh_path_ = "file://" + pkg_share + "/meshes/mecanum_robot.stl";
            } catch (...) {
                mesh_path_ = "";
            }
        }

        // Create TF broadcaster
        tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

        // Create publishers
        path_pub_ = this->create_publisher<nav_msgs::msg::Path>("/odometry/path", 10);
        pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("/odometry/pose", 10);
        marker_pub_ = this->create_publisher<visualization_msgs::msg::Marker>("/odometry/robot_marker", 10);
        label_pub_ = this->create_publisher<visualization_msgs::msg::MarkerArray>("/odometry/position_labels", 10);
        status_label_pub_ = this->create_publisher<visualization_msgs::msg::Marker>("/odometry/status_label", 10);

        // Create subscriber for position
        position_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odometry/position_from_accel",
            10,
            std::bind(&PositionVisualizer::position_callback, this, std::placeholders::_1));

        // Create subscriber for segment info
        segment_sub_ = this->create_subscription<std_msgs::msg::Int32MultiArray>(
            "/simulator/segment_info",
            10,
            std::bind(&PositionVisualizer::segment_callback, this, std::placeholders::_1));

        // Initialize path message
        path_msg_.header.frame_id = "map";

    // Start time will be initialized from the first incoming message's stamp
    start_time_initialized_ = false;

        RCLCPP_INFO(this->get_logger(), "Position Visualizer started");
        RCLCPP_INFO(this->get_logger(), "  Use mesh: %s", use_mesh_ ? "true" : "false");
        if (use_mesh_) {
            RCLCPP_INFO(this->get_logger(), "  Mesh path: %s", mesh_path_.c_str());
        }
    }

private:
    void position_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
    {
        // Get position and orientation
        double x = msg->pose.pose.position.x;
        double y = msg->pose.pose.position.y;
        double z = msg->pose.pose.position.z;
        
        // Get yaw from quaternion
        double qz = msg->pose.pose.orientation.z;
        double qw = msg->pose.pose.orientation.w;
        double yaw = 2.0 * std::atan2(qz, qw);

        // Skip NaN values (used for velocity-only updates)
        if (std::isnan(x) || std::isnan(y)) {
            return;
        }

        auto stamp = msg->header.stamp;

        // Publish TF: map -> base_link
        publish_transform(x, y, z, qz, qw, stamp);

        // Publish pose
        publish_pose(x, y, z, qz, qw, stamp);

        // Add to path
        add_to_path(x, y, z, qz, qw, stamp);

        // Publish robot marker
        publish_robot_marker(x, y, z, qz, qw, stamp);

        // Update status tracking (speed, distance, accel)
        update_status_tracking(x, y, msg->twist.twist.linear.x, msg->twist.twist.linear.y, stamp);

        // Publish floating status label above robot
        publish_floating_status(x, y, z, stamp);

        // Publish periodic labels (every second)
        // Check if 1 second has passed since last label
        // Calculate elapsed time since the first incoming message
        rclcpp::Time stamp_time(stamp);
        if (!start_time_initialized_) {
            start_time_ = stamp_time;
            start_time_initialized_ = true;
        }
        double elapsed_time = (stamp_time - start_time_).seconds();
        
        if (elapsed_time - last_label_time_ >= 1.0) {
            publish_time_label(x, y, z, elapsed_time, stamp_time);
            last_label_time_ = elapsed_time;
        }
    }

    void segment_callback(const std_msgs::msg::Int32MultiArray::SharedPtr msg)
    {
        if (msg->data.size() >= 2) {
            current_segment_ = msg->data[0];
            total_segments_ = msg->data[1];
        }
    }

    void publish_transform(double x, double y, double z, 
                          double qz, double qw, 
                          const rclcpp::Time& stamp)
    {
        geometry_msgs::msg::TransformStamped t;
        t.header.stamp = stamp;
        t.header.frame_id = "map";
        t.child_frame_id = "base_link";
        
        t.transform.translation.x = x;
        t.transform.translation.y = y;
        t.transform.translation.z = z;
        
        t.transform.rotation.x = 0.0;
        t.transform.rotation.y = 0.0;
        t.transform.rotation.z = qz;
        t.transform.rotation.w = qw;

        tf_broadcaster_->sendTransform(t);
    }

    void publish_pose(double x, double y, double z, 
                     double qz, double qw,
                     const rclcpp::Time& stamp)
    {
        geometry_msgs::msg::PoseStamped pose;
        pose.header.stamp = stamp;
        pose.header.frame_id = "map";
        
        pose.pose.position.x = x;
        pose.pose.position.y = y;
        pose.pose.position.z = z;
        
        pose.pose.orientation.x = 0.0;
        pose.pose.orientation.y = 0.0;
        pose.pose.orientation.z = qz;
        pose.pose.orientation.w = qw;

        pose_pub_->publish(pose);
    }

    void add_to_path(double x, double y, double z, 
                    double qz, double qw,
                    const rclcpp::Time& stamp)
    {
        geometry_msgs::msg::PoseStamped pose;
        pose.header.stamp = stamp;
        pose.header.frame_id = "map";
        
        pose.pose.position.x = x;
        pose.pose.position.y = y;
        pose.pose.position.z = z;
        
        pose.pose.orientation.x = 0.0;
        pose.pose.orientation.y = 0.0;
        pose.pose.orientation.z = qz;
        pose.pose.orientation.w = qw;

        path_msg_.poses.push_back(pose);
        
        // Limit path length
        if (static_cast<int>(path_msg_.poses.size()) > max_path_points_) {
            path_msg_.poses.erase(path_msg_.poses.begin());
        }

        path_msg_.header.stamp = stamp;
        path_pub_->publish(path_msg_);
    }

    // Helper to convert RPY to Quaternion
    struct Quaternion {
        double x, y, z, w;
    };

    Quaternion rpy_to_quaternion(double roll, double pitch, double yaw)
    {
        double cr = cos(roll * 0.5);
        double sr = sin(roll * 0.5);
        double cp = cos(pitch * 0.5);
        double sp = sin(pitch * 0.5);
        double cy = cos(yaw * 0.5);
        double sy = sin(yaw * 0.5);

        Quaternion q;
        q.w = cr * cp * cy + sr * sp * sy;
        q.x = sr * cp * cy - cr * sp * sy;
        q.y = cr * sp * cy + sr * cp * sy;
        q.z = cr * cp * sy - sr * sp * cy;
        return q;
    }

    void publish_robot_marker(double x, double y, double z, 
                             double qz, double qw,
                             const rclcpp::Time& stamp)
    {
        // 1. Base Chassis Marker (Cube)
        visualization_msgs::msg::Marker base_marker;
        base_marker.header.stamp = stamp;
        base_marker.header.frame_id = "base_link"; 
        base_marker.ns = "robot_base";
        base_marker.id = 0;
        base_marker.type = visualization_msgs::msg::Marker::CUBE;
        base_marker.action = visualization_msgs::msg::Marker::ADD;
        base_marker.frame_locked = true;

        base_marker.scale.x = 0.40;
        base_marker.scale.y = 0.20;
        base_marker.scale.z = 0.10;

        base_marker.pose.position.x = 0.0;
        base_marker.pose.position.y = 0.0;
        base_marker.pose.position.z = 0.05;
        base_marker.pose.orientation.w = 1.0;

        base_marker.color.r = 0.0f;
        base_marker.color.g = 0.0f;
        base_marker.color.b = 0.5f;
        base_marker.color.a = 0.9f;

        marker_pub_->publish(base_marker);

        // Wheels are rendered via URDF RobotModel in RViz - no need for separate markers
    }

    void publish_time_label(double x, double y, double z, double /*time_sec*/, const rclcpp::Time & stamp)
    {
        visualization_msgs::msg::MarkerArray markers;
        
        // Sphere marker with segment-based color
        visualization_msgs::msg::Marker sphere;
        sphere.header.frame_id = "map";
        sphere.header.stamp = stamp;
        sphere.ns = "position_spheres";
        sphere.id = label_id_++;
        sphere.type = visualization_msgs::msg::Marker::SPHERE;
        sphere.action = visualization_msgs::msg::Marker::ADD;
        sphere.pose.position.x = x;
        sphere.pose.position.y = y;
        sphere.pose.position.z = z;
        sphere.pose.orientation.w = 1.0;
        sphere.scale.x = 0.1;
        sphere.scale.y = 0.1;
        sphere.scale.z = 0.1;
        
        // Color based on current segment (different color per segment)
        set_sphere_color_by_segment(sphere);
        
        markers.markers.push_back(sphere);

        // No text labels - removed per user request 
        // (floating status label above robot shows all info now)

        label_pub_->publish(markers);
    }

    // Helper to set sphere color based on current segment
    void set_sphere_color_by_segment(visualization_msgs::msg::Marker& sphere)
    {
        // Color palette for segments (up to 10 distinct colors)
        const std::vector<std::array<float, 3>> segment_colors = {
            {1.0f, 0.0f, 0.0f},    // Segment 1: Red
            {0.0f, 1.0f, 0.0f},    // Segment 2: Green
            {0.0f, 0.5f, 1.0f},    // Segment 3: Blue
            {1.0f, 1.0f, 0.0f},    // Segment 4: Yellow
            {1.0f, 0.0f, 1.0f},    // Segment 5: Magenta
            {0.0f, 1.0f, 1.0f},    // Segment 6: Cyan
            {1.0f, 0.5f, 0.0f},    // Segment 7: Orange
            {0.5f, 0.0f, 1.0f},    // Segment 8: Purple
            {0.0f, 1.0f, 0.5f},    // Segment 9: Teal
            {1.0f, 0.0f, 0.5f}     // Segment 10: Pink
        };
        
        // Get color index (0-indexed, wrap around if more than 10 segments)
        int color_idx = (current_segment_ - 1) % segment_colors.size();
        
        sphere.color.r = segment_colors[color_idx][0];
        sphere.color.g = segment_colors[color_idx][1];
        sphere.color.b = segment_colors[color_idx][2];
        sphere.color.a = 1.0f;
    }

    void update_status_tracking(double x, double y, double vx, double vy, 
                                const builtin_interfaces::msg::Time& stamp)
    {
        rclcpp::Time current_time(stamp);
        
        if (prev_initialized_) {
            // Calculate distance increment
            double dx = x - prev_x_;
            double dy = y - prev_y_;
            double dist_increment = std::sqrt(dx * dx + dy * dy);
            total_distance_ += dist_increment;
            
            // Calculate current speed from velocity components
            current_speed_ = std::sqrt(vx * vx + vy * vy);
            
            // Calculate acceleration from velocity change
            double dt = (current_time - prev_time_).seconds();
            if (dt > 0.0) {
                double dvx = vx - prev_vx_;
                double dvy = vy - prev_vy_;
                double accel_x = dvx / dt;
                double accel_y = dvy / dt;
                current_accel_ = std::sqrt(accel_x * accel_x + accel_y * accel_y);
            }
            
            prev_vx_ = vx;
            prev_vy_ = vy;
        }
        
        prev_x_ = x;
        prev_y_ = y;
        prev_time_ = current_time;
        prev_initialized_ = true;
    }

    void publish_floating_status(double x, double y, double z, 
                                 const builtin_interfaces::msg::Time& stamp)
    {
        visualization_msgs::msg::Marker status;
        status.header.frame_id = "map";
        status.header.stamp = stamp;
        status.ns = "status_label";
        status.id = 0;
        status.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
        status.action = visualization_msgs::msg::Marker::ADD;
        
        // Position above the robot (higher than time labels)
        status.pose.position.x = x;
        status.pose.position.y = y;
        status.pose.position.z = z + 0.40;
        status.pose.orientation.w = 1.0;
        
        // Text size
        status.scale.z = 0.08;
        
        // White text
        status.color.r = 1.0f;
        status.color.g = 1.0f;
        status.color.b = 1.0f;
        status.color.a = 1.0f;
        
        // Calculate elapsed time
        double elapsed_time = 0.0;
        if (start_time_initialized_) {
            rclcpp::Time current_time(stamp);
            elapsed_time = (current_time - start_time_).seconds();
        }
        
        // Format the multi-line status text like in the screenshot
        std::stringstream ss;
        ss << std::fixed;
        ss << "seg:" << current_segment_ << "/" << total_segments_ << "\n";
        ss << "time:" << std::setprecision(0) << elapsed_time << "s\n";
        ss << "dist:" << std::setprecision(2) << total_distance_ << "m\n";
        ss << "speed:" << std::setprecision(2) << current_speed_ << "m/s\n";
        ss << "accel:" << std::setprecision(2) << current_accel_ << "m/s^2";
        
        status.text = ss.str();
        
        status_label_pub_->publish(status);
    }

    // Parameters
    bool use_mesh_;
    double mesh_scale_;
    std::string mesh_path_;
    double path_width_;
    int max_path_points_;

    // Path storage
    nav_msgs::msg::Path path_msg_;
    
    // Label tracking
    double last_label_time_{0.0};
    int label_id_{0};
    rclcpp::Time start_time_;              // Start time (initialized from first message)
    bool start_time_initialized_{false};  // True after first incoming message stamp is stored

    // Status tracking for floating label
    double total_distance_{0.0};           // Total distance traveled (meters)
    double current_speed_{0.0};            // Current speed (m/s)
    double current_accel_{0.0};            // Current acceleration magnitude (m/s^2)
    double prev_x_{0.0}, prev_y_{0.0};     // Previous position for distance calculation
    double prev_vx_{0.0}, prev_vy_{0.0};   // Previous velocity for acceleration calculation
    rclcpp::Time prev_time_;               // Previous timestamp
    bool prev_initialized_{false};         // True after first position received
    int current_segment_{1};               // Current segment number
    int total_segments_{5};                // Total segments (will be updated from parameter)
    
    // Publisher for floating status label
    rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr status_label_pub_;

    // TF broadcaster
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;

    // Publishers
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub_;
    rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_pub_;
    rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr marker_pub_;
    rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr label_pub_;

    // Subscriber
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr position_sub_;
    rclcpp::Subscription<std_msgs::msg::Int32MultiArray>::SharedPtr segment_sub_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PositionVisualizer>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
